Feature/values pattern

dpetran opened the pull request:
This adds support for `:values` as a where-clause pattern. This allows users to scope their `:values` to just a `:union` or `:optional` pattern, instead of constraining the whole `:where` clause.

This will enable us to translate the SPARQL VALUES pattern directly to FQL.

dpetran committed:
add support for multiple values per var in VALUES pattern

In SPARQL, a VALUES pattern can appear inside a WHERE clause alongside other
patterns. In FQL :values is its own clause. This mismatch makes translating directly
awkward, since we need to insert our translated pattern at a different point in the
parse tree than where we are at translation time.

We got around it by translating it into a :bind pattern. However, a :bind pattern can
only bind a single value to a variable, so we limited the scope of a VALUES pattern to
only a single value.

This gets around it by "tagging" the translated value pattern and then after the whole
parsing is done, moving it to the correct place in the tree: see `format-values`. It's
awkward, but it gives much more flexibility to the user.

dpetran committed:
support VALUES patterns in where clause

This is a basic implementation of the SPARQL VALUES pattern type.

dpetran committed:
support VALUES patterns in an order-agnostic way

Three approaches considered:

1) just document that :values needs to appear first
2) discard solutions that don't match the given solution
3) during the parse step, sort the patterns so :values is first

I don't like #1, though that's the most expedient. #2 is a bit wasteful as it forces us
to generate useless solutions. #3 is doable, and may be a minimal requirement in basic
query planning.

This implements solution #2.

dpetran committed:
add values test for nested patterns

dpetran committed:
add support for VALUES patterns and VALUES clauses

FQL now supports :values patterns so we can use that directly instead of translating to
a :values clause.

dpetran committed:
SUBSTR and REPLACE expresssion translations

dpetran committed:
Throw errors for unsupported EXISTS and NOT EXISTS filters

dpetran committed:
add support for complex multiplicative expressions

dpetran committed:
handle Func translation on a case-by-case basis

The tricky part is deciding when to `literal-quote` one or more of the args. Since that
information does not exist in the parse tree at the point where :Func needs parsing, we
need to figure out whether to quote based on the signature of the function, according to
the SPARQL spec: https://www.w3.org/TR/sparql11-query/#SparqlOps and the SPARQL grammar.

This way is more repetitive but more robust. Once we've got all the functions covered we
can search for abstractions to reduce verbosity.

dpetran committed:
add support for BOUND

JaceRockman committed:
Add separate test for funcs

JaceRockman committed:
Alphabetize the lists of funcs

dpetran committed:
support SPARQL literal datatype and language tags

I'm not sure where the LANGTAG regex came from, but it didn't match the one in the spec
- it required the hyphen (-) between two-part lang tags of the style `@en-us`, but
disallowed the simple verson `@en`. Now it correctly accepts both.

dpetran committed:
confirm SPARQL comments are ignored

JaceRockman committed:
add support for BNODE

JaceRockman committed:
add support for COALESCE

JaceRockman committed:
fix implementation of BNODE

JaceRockman committed:
add support for CONTAINS

JaceRockman committed:
add support for DATATYPE

JaceRockman committed:
add support for DAY

JaceRockman committed:
add support for ENCODE_FOR_URI

JaceRockman committed:
add support for FLOOR

JaceRockman committed:
add support for HOURS

JaceRockman committed:
add support for IF

JaceRockman committed:
add support for IRI

JaceRockman committed:
add support for LANG

JaceRockman committed:
add support for LANGMATCHES

JaceRockman committed:
add support for LCASE

JaceRockman committed:
add support for MD5

JaceRockman committed:
add support for MINUTES

JaceRockman committed:
add support for MONTH

JaceRockman committed:
add support for NOW

JaceRockman committed:
add support for RAND

JaceRockman committed:
add support for ROUND

JaceRockman committed:
add support for SECONDS

JaceRockman committed:
add support for SHA1

JaceRockman committed:
add support for SHA256

JaceRockman committed:
add tests for SHA512

JaceRockman committed:
add support for STR

JaceRockman committed:
add support for STRAFTER and STRBEFORE

JaceRockman committed:
add support for STRDT

JaceRockman committed:
add support for STRENDS

JaceRockman committed:
add support for STRLANG

JaceRockman committed:
add support for STRLEN and STRENDS

JaceRockman committed:
add support for STRUUID

JaceRockman committed:
add support for TIMEZONE

JaceRockman committed:
add support for TZ

JaceRockman committed:
add support for UCASE

JaceRockman committed:
add support for URI

JaceRockman committed:
add support for UUID

JaceRockman committed:
add support for YEAR

JaceRockman committed:
add support for isBLANK

JaceRockman committed:
add support for isIRI

JaceRockman committed:
add support for isLITERAL

JaceRockman committed:
add support for isNUMERIC

JaceRockman committed:
add support for isURI

JaceRockman committed:
add support for sameTerm

dpetran committed:
correctly handle multiple expressions joined logically

dpetran committed:
add error for SERVICE patterns

JaceRockman committed:
add support for FROM NAMED

JaceRockman committed:
Correct grammar comment in DatasetClause parse-rule

JaceRockman committed:
add support for GRAPH

dpetran committed:
add support for EXISTS and NOT EXISTS patterns

dpetran committed:
use existing helper function instead of destructuring

dpetran committed:
add explanations to test cases

dpetran committed:
use parse-where-clause function

dpetran committed:
allow SPARQL to use `a` as an alias for @type

The grammar wasn't expecting whitespace between the 'a' and the prev/next term.

zonotope commented on src/clj/fluree/db/query/exec/where.cljc:
```diff
@@ -566,6 +566,23 @@
+                                       (into {} )))
+                                inline-solutions)]
+    ;; filter out any solutions that don't match inline solution data
+    (if (every? (fn [inline-solution] (= (select-keys solution (keys inline-solution))
```
I don't think this will work in all cases. There could be other metadata that we pull from the dbs that won't be in the inline solution. There might also be problems in the context of federated queries where a subject was pulled from a different database than the one being evaluated now, so the sid that you compute above won't match. I think you have to pull the value or iri out specifically instead of relying on the entire match map to be equal with the inline one for each inline variable.

dpetran replied:
Hmm, maybe `:lang` from :meta as well? I'm not sure about `:i`, though.

zonotope commented on src/clj/fluree/db/query/fql/parse.cljc:
```diff
@@ -570,7 +575,7 @@
   [q]
   (let [context       (context/extract q)
-        [vars values] (parse-values q context)
+        [vars values] (parse-values (:values q) context)
```
Was there a problem with passing the entire query to the `parse-values` function? I would prefer to follow that pattern because there are some steps which might need to get information from multiple parts of the query, and I think the parsing code is clearer when each parsing function acts on the query consistently.

dpetran replied:
Not so much a problem as it being more inconvenient. `parse-values` is called in three places:
- `parse-analytical-query`, where passing the entire query is easy
- `parse-txn`, where the values have to be put into an artificially constructed query map
- `parse-pattern :values`, where we would have to artificially construct a query map

On the balance of uses, I thought it was cleaner to refactor it to just accept the values instead of assuming that it comes attached to a query, since most of the time it's actually not.

dpetran committed:
correctly filter solutions across ledgers and with meta

Change 1: handle multiple inline values correctly.

Instead of checking that _every_ inline solution equal the solution matches, we needed
to _filter_ the inline solutions by whether they match the solution. If any matches, the
values can contribute their solutions to the solution chan.

This case allows multiple values to be handled correctly, as seen in the "pattern"
"single var" test.

Change 2: do not consider ::sids while checking match equality

When dealing with federated queries the ::sids map may be different for the same
value. We now dissoc the ::sids key from the solution matches before checking equality.

This is verified by the "federated" test case

Change 3: only consider :lang ::meta while checking equality

Two identical strings with different language tags are different values. However the :i
order or whether it comes from a reasoned flake do not matter for equality checks. We
now select only the :lang ::meta key, if ::meta exists in the solution match.

Language tags were not parsed, so I've added parsing support for that.

zonotope commented on src/clj/fluree/db/query/exec/where.cljc:
```diff
@@ -566,6 +566,22 @@
+  (let [inline-solutions (pattern-data pattern)
+        ;; need to remove ::sids from the solution as those may differ based on which graph they are from
+        solution* (reduce (fn [soln [var match]]
+                            (assoc soln var (cond-> (dissoc match ::sids)
```
I think this is still brittle and could unintentionally break if we add more information to the matches. Instead of removing known divergent information from the solution and doing a top level equality check, I think we should reduce over the variables in the inline solution and explicitly check if the value and datatype of the match bound to that variable in the solution is the same as the value and datatype of the inline solution and, if there is no value/datatype, we should check if it's a matched iri and the iris are the same. The `get-value` and `get-datatype` functions will help you in the first case, and the `get-iri` will help you in the second. 

We should explicitly check the things we know must be the same, instead of explicitly eliminating the things we think might be different, because we could add more things that might be different later.

dpetran replied:
Good point, I've updated this section (and discovered the very handy `update-vals` core fn, thanks Rich) to select the important bits.

dpetran committed:
use public match api instead of removing unwanted match elements

In the future the contents of a match may change, so instead of removing specific
unwanted elements we now extract only those elements of a match that establish its
identity for equality checking.

dpetran committed:
fix typo

dpetran committed:
add description of sparql function translation

dpetran committed:
add support for `in` expression

Semantics and syntax from SPARQL IN expression.

dpetran committed:
translate SPARQL EXISTS/NOT EXISTS filters into FQL

dpetran committed:
SPARQL IN support

dpetran committed:
fix expression quoting

The :Expression parser was turning everything into a string, which works in some cases
but introduces unwanted quoting in others. By deferring stringification until values
need to be incorporated into an actual string expression we avoid the unwanted quoting.

Do note that all the numeric functions now correctly expect numeric values instead of
strings.

dpetran committed:
SPARQL NOT IN support

dpetran committed:
correctly assign langString datatype to language tagged strings

We were not parsing language tagged strings to langString datatypes correctly in where
patterns or in values patterns, nor inferring a langString datatype in the presence of a
language tag.

I had to update the test expectation because now the language tagged string no longer
conforms to the `sh:maxCount 1` _and_ the `sh:datatype xsd:string` constraints, instead of
just the `sh:maxCount` constraint.

dpetran commented:
Rebased on https://github.com/fluree/db/pull/765

dpetran committed:
remove useless thread macro

dpetran committed:
rename sparql-in to in

It's not sparql-specific, so no need to qualify it.

dpetran committed:
fix abs expression parsing

abs expects only a single expression as an argument, and that argument does not need to
be quoted.

dpetran committed:
no need to literal-quote numeric functions

dpetran committed:
add more explanatory detail to test forms

dpetran committed:
don't clobber existing solution bindings with inline values

zonotope reviewed (approved):
🥼

dpetran committed:
Merge pull request #797 from fluree/feature/sparql-in-not-in

translate SPARQL IN/NOT IN expresssions into FQL

dpetran committed:
Merge pull request #796 from fluree/feature/sparql-exists-not-exists

translate SPARQL EXISTS/NOT EXISTS filters into FQL

dpetran committed:
Merge pull request #795 from fluree/feature/fql-in-not-in

add support for `in` expression

dpetran committed:
Merge pull request #784 from fluree/feature/fql-exists-not-exists

Feature/fql exists not exists

dpetran committed:
Merge pull request #780 from fluree/feature/sparql-from-named

Feature/sparql from named

dpetran committed:
Merge pull request #778 from fluree/feature/sparql-func

Feature/sparql func

dpetran committed:
Merge branch 'feature/sparql-values' into feature/values-pattern

dpetran commented:
The CI was stuck waiting for a container so I merged and ran CI locally.
